File: cis_interface/examples/formatted_io5/src/formatted_io5_modelA.m
% Initialize input/output channels 
in_channel = CisInterface('CisPlyInput', 'inputA');
out_channel = CisInterface('CisPlyOutput', 'outputA');

flag = true;

% Loop until there is no longer input or the queues are closed
while flag

  % Receive input from input channel
  % If there is an error, the flag will be False.
  [flag, ply] = in_channel.recv();
  if (~flag)
    disp('Model A: No more input.');
    break;
  end;

  % Print received message
  fprintf('Model A: (%d verts, %d faces)\n', ply.nvert, ply.nface);
  fprintf('  Vertices:\n');
  for i = 1:int64(ply.nvert)
    fprintf('   %f, %f, %f\n', ...
            ply{'vertices'}{i}{'x'}, ply{'vertices'}{i}{'y'}, ply{'vertices'}{i}{'z'});
  end;
  fprintf('  Faces:\n');
  for i = 1:int64(ply.nface)
    fprintf('   %d', ply{'faces'}{i}{'vertex_index'}{1});
    for j = 2:size(ply{'faces'}{i}{'vertex_index'})
      fprintf(', %d', ply{'faces'}{i}{'vertex_index'}{j});
    end;
    fprintf('\n');
  end;

  % Send output to output channel
  % If there is an error, the flag will be False
  flag = out_channel.send(ply);
  if (~flag)
    error('Model A: Error sending output.');
    break;
  end;
  
end;

File: yggdrasil/examples/formatted_io5/src/formatted_io5_modelA.f90
program main
  ! Include methods for input/output channels
  use fygg

  ! Declare resulting variables and create buffer for received message
  logical :: flag = .true.
  type(yggcomm) :: in_channel, out_channel
  type(yggply) :: p
  p = init_ply()

  ! initialize input/output channels
  in_channel = ygg_ply_input("inputA")
  out_channel = ygg_ply_output("outputA")

  ! Loop until there is no longer input or the queues are closed
  do while (flag)

     ! Receive input from input channel
     ! If there is an error, the flag will be negative
     ! Otherwise, it is the number of variables filled
     flag = ygg_recv_var(in_channel, yggarg(p))
     if (.not.flag) then
        print *, "Model A: No more input."
        exit
     end if

     ! Print received message
     print *, "Model A: (", p%nvert, " verts, ", p%nface, " faces)"
     call display_ply_indent(p, "  ")

     ! Send output to output channel
     ! If there is an error, the flag will be negative
     flag = ygg_send_var(out_channel, yggarg(p))
     if (.not.flag) then
        print *, "Model A: Error sending output."
        exit
     end if

  end do

end program main
